src/components/validation/validators/component_spec.rs

```diff
@@ -23,11 +23,60 @@ impl Validator for ComponentSpecValidator {
     fn check_validation(
         &self,
         _component_type: ComponentType,
-        _expectation: TestCaseExpectation,
+        expectation: TestCaseExpectation,
         inputs: &[TestEvent],
         outputs: &[Event],
         telemetry_events: &[Event],
     ) -> Result<Vec<String>, Vec<String>> {
+        for input in inputs {
+            debug!("Validator observed input event: {:?}", input);
+        }
+
+        for output in outputs {
+            debug!("Validator observed output event: {:?}", output);
+        }
+
+        // Validate that the number of inputs/outputs matched the test case expectation.
+        match expectation {
+            TestCaseExpectation::Success => {
+                if inputs.len() != outputs.len() {
+                    return Err(vec![format!(
+                        "sent {} inputs but only received {} outputs",
+                        inputs.len(),
+                        outputs.len()
+                    )]);
+                }
+            }
+            TestCaseExpectation::Failure => {
+                if !outputs.is_empty() {
+                    return Err(vec![format!(
+                        "received {} outputs but none were expected",
+                        outputs.len()
+                    )]);
+                }
+            }
+            TestCaseExpectation::PartialSuccess => {
+                // TODO: This one is a bit loosy goosy because we should really be checking if we
+                // got an output event for every valid input event, where, ostensibly, any event
+                // that wasn't marked for modification would represent a valid input event.
+                //
+                // Thinking it through, however, it's not 100% clear at this moment whether or not
+                // we'll ever have a test case, with an expectation of partial success, that
+                // intentionally modifies one or more input events but where some of those modified
+                // events will lead to output events.
+                //
+                // I might be overthinking it, but we'll leave this as-is for now because it's
+                // certainly the simplest check that matches reality.
```

Comment:
YAGNI approach sounds good here, thanks for calling it out in detail.

Is this really a TODO though? 💭

Reply:
Yeah, I'll transform it to a more succinct note somewhere above.

Reply:
Fixed in 386072039.